File: skills/code-architect/references/code-examples/typescript/Proxy.ts
// Proxy — a placeholder controlling access to a real object (lazy + caching).

interface ImageResource {
  fetch(): string;
}

// The real, expensive subject.
class RemoteImage implements ImageResource {
  constructor(private readonly url: string) {
    console.log(`loading heavy resource: ${url}`);
  }
  fetch(): string {
    return `<bytes of ${this.url}>`;
  }
}

// Proxy defers creation and caches the result.
class LazyImageProxy implements ImageResource {
  private real: RemoteImage | null = null;
  private cached: string | null = null;

  constructor(private readonly url: string) {}

  fetch(): string {
    if (this.real === null) {
      this.real = new RemoteImage(this.url);
    }
    if (this.cached === null) {
      this.cached = this.real.fetch();
    }
    return this.cached;
  }
}

function demo(): void {
  const image: ImageResource = new LazyImageProxy("photo.png");
  console.log("proxy created (nothing loaded yet)");
  console.log(image.fetch()); // triggers load
  console.log(image.fetch()); // served from cache
}

demo();

File: skills/code-architect/references/code-examples/ruby/proxy.rb
# frozen_string_literal: true

# Proxy — a placeholder controlling access to a real object (lazy + caching here).

# Real subject: expensive to create, expensive to query.
class RemoteResource
  def initialize(name)
    @name = name
    puts "  (loaded #{name} from network)"
  end

  def fetch(key) = "#{@name}:#{key}=#{key.length}"
end

# Proxy shares the subject interface, defers creation, and caches results.
class ResourceProxy
  def initialize(name) = @name = name

  def fetch(key)
    @cache ||= {}
    @cache.fetch(key) do
      @cache[key] = subject.fetch(key)
    end
  end

  private

  def subject
    @subject ||= RemoteResource.new(@name) # lazy: built on first real use
  end
end

if __FILE__ == $PROGRAM_NAME
  proxy = ResourceProxy.new("catalog")
  puts "proxy created (nothing loaded yet)"

  puts proxy.fetch("alpha") # triggers load
  puts proxy.fetch("alpha") # served from cache, no reload
  puts proxy.fetch("beta")  # cache miss, subject already loaded
end
